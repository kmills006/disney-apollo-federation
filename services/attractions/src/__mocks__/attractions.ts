export const MOCK_ATTRACTIONS = [
  {
    name: 'Attraction One',
    short_name: 'one',
    permalink: 'attraction-one',
  },
  {
    name: 'Attraction Two',
    short_name: 'two',
    permalink: 'attraction-two',
  },
];
