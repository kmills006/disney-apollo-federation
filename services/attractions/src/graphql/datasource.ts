import { RESTDataSource } from 'apollo-datasource-rest';
import { InMemoryLRUCache } from 'apollo-server-caching';

export enum ParkPermalink {
  MagicKingdom = 'magic-kingdom',
  Epcot = 'epcot',
  AnimalKingdom = 'animal-kingdom',
  HollywoodStudios = 'hollywood-studios',
}

export class AttractionAPI extends RESTDataSource {
  private endpoint = 'attractions.json';

  constructor() {
    super();

    this.baseURL = 'https://touringplans.com/';

    // Required in order to use the RESTDatasource without passing it
    // to an ApolloServer
    const cache = new InMemoryLRUCache();
    this.initialize({ context: {}, cache });
  }

  public async getAttractions(park: ParkPermalink) {
    console.log('parkPermalinks', park);

    return this.get(`${park}/${this.endpoint}`);
  }
}
