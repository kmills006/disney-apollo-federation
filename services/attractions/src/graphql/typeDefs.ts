import { gql } from 'apollo-server';

export const typeDefs = gql`
  type Query {
    attractions: [Attraction!]
  }

  type Attraction @key(fields: "permalink") {
    name: String!
    permalink: String!
  }

  extend type Park @key(fields: "permalink") {
    permalink: String! @external

    attractions: [Attraction!]
  }
`;
