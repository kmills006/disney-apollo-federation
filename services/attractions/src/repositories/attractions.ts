import * as A from 'fp-ts/lib/Array';
import { pipe } from 'fp-ts/lib/pipeable';
import * as TE from 'fp-ts/lib/TaskEither';

import { AttractionAPI, ParkPermalink } from '../graphql/datasource';
import { IAttraction } from '../models/Attraction';

export interface IAttractionsRepository {
  getAttractions: () => TE.TaskEither<Error, IAttraction[]>;

  getParkAttractions: (park:ParkPermalink) => TE.TaskEither<Error, IAttraction[]>;
}

export const repository = (api: AttractionAPI) => (): IAttractionsRepository => ({
  getAttractions() {
    const ops = Object.values(ParkPermalink).map((park) => this.getParkAttractions(park));

    return pipe(
      A.array.sequence(TE.taskEither)(ops),
      TE.map((a) => a.flat()),
    );
  },

  getParkAttractions: (park) => TE.tryCatch(
    () => api.getAttractions(park),
    (reason) => reason as Error,
  ),
});

export const attractionsRepository = repository(new AttractionAPI());
