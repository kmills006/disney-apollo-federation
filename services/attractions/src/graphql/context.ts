import { attractionsRepository, IAttractionsRepository } from '../repositories/attractions';

export interface IResolverContext {
  repositories: {
    attractions: IAttractionsRepository,
  },
}

export const context = (): IResolverContext => ({
  repositories: {
    attractions: attractionsRepository(),
  },
});
