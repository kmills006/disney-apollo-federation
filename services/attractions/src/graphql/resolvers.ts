import { ApolloError } from 'apollo-server';
import { ResolverFunc } from '@disney-federation/gql-utils';
import * as T from 'fp-ts/lib/Task';
import * as TE from 'fp-ts/lib/TaskEither';
import * as P from 'fp-ts/lib/pipeable';

import { IResolverContext } from './context';
import { Attraction } from '../models/Attraction';
import { AttractionDetail } from '../models/AttractionDetail';
import { Park } from '../models/Park';

export interface IQueryResolvers {
  attractions: ResolverFunc<Promise<Attraction[]>, IResolverContext>;
}

type ParkResolverFunc =
  ResolverFunc<Promise<Attraction[]>, IResolverContext, never, Park>;

export interface IParkResolvers {
  attractions: ParkResolverFunc;
}

type IAttractionDetailResolver = ResolverFunc<any, IResolverContext, never, Attraction>;

export interface IAttractionResolvers {
  detail: IAttractionDetailResolver;
}

export interface IResolvers {
  [key: string]: {};
  Attraction: IAttractionResolvers;
  Query: IQueryResolvers;
  Park: IParkResolvers;
}

// TODO: Make a proper Error object instead of throwing
const handleError = (e: Error) => { throw new ApolloError(e.message); };
const handleResponse = <T>(a: T) => T.of(a);

export const resolvers: IResolvers = {
  Attraction: {
    detail: async (parent, _, ctx) => P.pipe(
      ctx.repositories.attractions.getParkAttractionDetails(
        parent.park.permalink,
        parent.permalink,
      ),
      TE.fold(handleError, (r) => handleResponse<AttractionDetail>(r)),
    )(),
  },

  Park: {
    attractions: (parent, _, ctx) => P.pipe(
      ctx.repositories.attractions.getParkAttractions(parent.permalink),
      TE.fold(handleError, (r) => handleResponse<Attraction[]>(r)),
    )(),
  },

  Query: {
    attractions: async (_, __, ctx) => P.pipe(
      ctx.repositories.attractions.getAttractions(),
      TE.fold(handleError, (r) => handleResponse<Attraction[]>(r)),
    )(),
  },
};
