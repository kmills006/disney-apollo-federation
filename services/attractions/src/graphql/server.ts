import { ApolloServer } from 'apollo-server';
import { buildFederatedSchema } from '@apollo/federation';

import { AttractionAPI } from './datasource';
import { attractionResolvers } from './resolvers';
import { typeDefs } from './typeDefs';

const schema = buildFederatedSchema({
  typeDefs,
  resolvers: attractionResolvers,
});

export const initiateApolloServer = () =>
  new ApolloServer({
    schema,
    dataSources: () => ({
      attractions: new AttractionAPI(),
    }),
  });
