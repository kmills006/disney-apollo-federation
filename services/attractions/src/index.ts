import { startApolloServer } from '@disney-federation/gql-utils';

import { server } from './graphql/server';

startApolloServer(4001)(server).catch((error) => {
  console.error(`Unable to start Attractions server: ${error.message}`);
  process.exit(1);
});
