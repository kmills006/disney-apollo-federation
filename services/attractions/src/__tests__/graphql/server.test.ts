import { buildFederatedSchema } from '@apollo/federation';
import { constructApolloServer } from '@disney-federation/gql-utils';
import { ApolloServer, gql } from 'apollo-server';
import { createTestClient } from 'apollo-server-testing';
import { mockDeep } from 'jest-mock-extended';
import * as TE from 'fp-ts/lib/TaskEither';

import { MOCK_ATTRACTIONS } from '../../__mocks__/attractions';
import { resolvers } from '../../graphql/resolvers';
import { typeDefs } from '../../graphql/typeDefs';
import { IAttractionsRepository } from '../../repositories/attractions';

describe('graphql', () => {
  describe('server', () => {
    const mockAttractionsRepository = mockDeep<IAttractionsRepository>();

    let server: ApolloServer;

    beforeEach(() => {
      server = constructApolloServer({
        schema: buildFederatedSchema({ resolvers, typeDefs }),
        context: () => ({
          repositories: { attractions: mockAttractionsRepository },
        }),
      });
    });

    describe('Park', () => {

    });

    describe('queries', () => {
      describe('attractions', () => {
        const GET_ATTRACTIONS = gql`query { attractions { name permalink } }`;

        it('returns a list of attractions', async () => {
          mockAttractionsRepository.getAttractions.mockReturnValue(TE.right(MOCK_ATTRACTIONS));

          const { query } = createTestClient(server);
          const actual = await query({ query: GET_ATTRACTIONS });

          expect(actual).toMatchSnapshot();
        });

        it('throws an ApolloError', async () => {
          mockAttractionsRepository.getAttractions.mockReturnValue(TE.left(new Error('forced error')));

          const { query } = createTestClient(server);
          const actual = await query({ query: GET_ATTRACTIONS });

          expect(actual).toMatchSnapshot();
        });
      });
    });
  });
});
