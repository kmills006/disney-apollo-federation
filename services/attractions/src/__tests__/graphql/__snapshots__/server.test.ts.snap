// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql server queries attractions returns a list of attractions 1`] = `
Object {
  "data": Object {
    "attractions": Array [
      Object {
        "name": "Attraction One",
        "permalink": "attraction-one",
      },
      Object {
        "name": "Attraction Two",
        "permalink": "attraction-two",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`graphql server queries attractions throws an ApolloError 1`] = `
Object {
  "data": Object {
    "attractions": null,
  },
  "errors": Array [
    [GraphQLError: forced error],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
